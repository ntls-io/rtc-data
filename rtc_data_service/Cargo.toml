[package]
name = "rtc_data_service"
version = "0.0.1"
authors = ["Registree"]
edition = "2018"

[lib]
crate-type = ["rlib"]

[dependencies]
sgx_types = { git = "https://github.com/apache/teaclave-sgx-sdk.git", features = ["extra_traits"] }
sgx_urts = { git = "https://github.com/apache/teaclave-sgx-sdk.git", rev = "v1.1.3" }
mockall_double = { version = "0.2.0", features = ["nightly"] }
thiserror = "1.0.24"
base64 = "0.13.0"
actix-web = "4.0.0-beta.5"
actix = "0.11"
serde = { version = "1.0.125", features = ["derive"] }
config = "0.11.0"
rtc_uenclave = { path = "../rtc_uenclave" }
rtc_types = { path = "../rtc_types" }
serde_json = "1.0.64"
actix-rt = "2.2.0"
cfg-if = "1.0.0"

[dev-dependencies]
rand = "0.7.3"
mockall = { version = "0.9.1", features = ["nightly"] }
proptest = "1.0.0"
num-bigint = { version = "0.4.0", features = ["rand"] }
num-traits = "0.2"
# This should be removed once the rust-crypto rsa crate updates with export functionality
simple_asn1 = { version = "0.5.1" }
insta = "1.7.1"
sodalite = "0.4.0"

[features]
test = []

[build-dependencies]
cc = "1.0.67"
