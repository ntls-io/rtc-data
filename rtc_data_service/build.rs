use std::env;

fn main() {
    let test_enabled = env::var_os("CARGO_FEATURE_TEST").is_some();

    let cur_dir = env::current_dir().unwrap();

    let sdk_dir = env::var("SGX_SDK").unwrap_or_else(|_| "/opt/sgxsdk".to_string());
    let is_sim = env::var("SGX_MODE").unwrap_or_else(|_| "HW".to_string());

    // This check allows unit tests to run without the need to link and libraries
    // by using `cargo test --features=test`
    if !test_enabled {
        println!(
            "cargo:rustc-link-search=native={}/build/data_system/lib",
            cur_dir.parent().unwrap().to_str().unwrap()
        );

        println!("cargo:rustc-link-lib=static=Enclave_u");

        println!("cargo:rustc-link-search=native={}/lib64", sdk_dir);
        println!("cargo:rustc-link-lib=static=sgx_uprotected_fs");

        println!("cargo:rustc-link-lib=dylib=sgx_dcap_ql");
        match is_sim.as_ref() {
            "SW" => {
                println!("cargo:rustc-cfg=sgx_mode=\"SW\"");
                println!("cargo:rustc-link-lib=dylib=sgx_urts_sim");
                println!("cargo:rustc-link-lib=dylib=sgx_uae_service_sim");
            }
            _ => {
                // HW by default
                println!("cargo:rustc-cfg=sgx_mode=\"HW\"");
                println!("cargo:rustc-link-lib=dylib=sgx_urts");
                println!("cargo:rustc-link-lib=dylib=sgx_uae_service");
            }
        }
    }
}
