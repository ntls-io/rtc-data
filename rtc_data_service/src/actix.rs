mod actix_config{
    use config::ConfigError;
    use serde::Deserialize;

    // Configuration specific to the server
    #[derive(Deserialize)]
    pub struct ServerConfig {
        pub host : String,
        pub port : i32
    }


    // App configuration
    #[derive(Deserialize)]
    pub struct Config {
        pub server : ServerConfig
    }

    impl Config {
        // Loads configuration from the environment to the Config Struct
        pub fn from_env() -> Result<Self, ConfigError> {
            let mut cfg = config::Config::new();
            cfg.merge(config::Environment::new())?;
            cfg.try_into()
        }
    }
}

pub mod actix_server{
    use actix_web::{App, HttpResponse, HttpServer, Responder, web};
    use crate::actix::models::Status;
    use dotenv::dotenv;
    
    pub async fn server_status() -> impl Responder {
        HttpResponse::Ok()
            .json(Status {status : "The server is up".to_string()})
    }
    
    #[actix_web::main]
    pub async fn start_server() -> std::io::Result<()> {
    
        dotenv().ok();
    
        let config = crate::actix::actix_config::Config::from_env().expect("Server config expected");
    
        print!("Starting server at http://{}:{}/", config.server.host, config.server.port);
        HttpServer::new(|| {
            App::new()
                .route("/", web::get().to(server_status))
        })
        .bind(format!("{}:{}", config.server.host, config.server.port))?
        .run()
        .await
    }
}

mod models{
    use serde::Serialize;

    #[derive(Serialize)]
    pub struct Status {
        pub status : String
    }
}
