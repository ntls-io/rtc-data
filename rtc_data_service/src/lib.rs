#![deny(clippy::mem_forget)]
#![feature(toowned_clone_into)]
#![feature(try_blocks)]
#![warn(rust_2018_idioms)]

pub mod app_config;
pub mod auth_enclave_actor;
pub mod data_enclave_actor;
pub mod data_upload;
pub mod exec_token;
pub mod handlers;
pub mod merge_error;
pub mod validation;

use base64;
use base64_serde::base64_serde_type;

base64_serde_type!(pub Base64Standard, base64::STANDARD);
