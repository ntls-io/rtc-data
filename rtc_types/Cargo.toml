[package]
name = "rtc_types"
version = "0.1.0"
authors = ["Herman <herman@registree.io>"]
edition = "2018"

[lib]
crate-type = ["staticlib", "rlib"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[features]
default = ["thiserror"]
teaclave_sgx = ["thiserror_sgx", "sgx_tstd"]

# See "Pinning SGX dependencies" in HACKING.md
[dependencies]
sgx_types = { git = "https://github.com/apache/incubator-teaclave-sgx-sdk.git", rev = "b9d1bda", features = ["extra_traits"] }

# no_std libraries
rkyv = { version = "0.6.6", default_features = false, features = ["const_generics", "strict"] }

# default
thiserror = { version = "1.0.24", optional = true}

# teaclave_sgx
thiserror_sgx = { git = "https://github.com/mesalock-linux/thiserror-sgx.git", tag = "sgx_1.1.3", package = "thiserror", optional = true }
sgx_tstd = { git = "https://github.com/apache/incubator-teaclave-sgx-sdk.git", rev = "b9d1bda", features = ["backtrace"], optional = true }

[dev-dependencies]
proptest = "1.0.0"
proptest-derive = "0.3.0"
