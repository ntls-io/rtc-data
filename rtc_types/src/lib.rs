// TODO: Document
#![cfg_attr(feature = "teaclave_sgx", no_std)]
#[cfg(feature = "teaclave_sgx")]
extern crate sgx_tstd as std;
extern crate sgx_types;
#[cfg(not(feature = "teaclave_sgx"))]
extern crate thiserror;
#[cfg(feature = "teaclave_sgx")]
extern crate thiserror_sgx as thiserror;

use std::fmt::Display;
use thiserror::Error;

use sgx_types::*;

use std::boxed::Box;

mod data_upload;
pub mod dh;
pub use data_upload::*;

mod exec_token;
pub use exec_token::*;

mod ecall_result;
pub use ecall_result::*;

#[repr(C)]
#[derive(Clone, Debug)]
pub struct EncryptedMessage {
    pub ciphertext: Box<[u8]>,
    pub nonce: [u8; 24],
}

#[repr(C)]
#[derive(Clone, Debug)]
pub struct SizedEncryptedMessage<const MESSAGE_LEN: usize> {
    pub ciphertext: [u8; MESSAGE_LEN],
    pub nonce: [u8; 24],
}

#[repr(C)]
#[derive(Copy, Clone, PartialEq, Eq, Ord, PartialOrd, Debug, Error)]
pub enum CryptoError {
    #[error("Crypto rng error: {}", .0)]
    Rand(u32),
    #[error("Unknown crypto error")]
    Unknown,
}

#[repr(C)]
#[derive(Copy, Clone, PartialEq, Eq, Ord, PartialOrd, Debug, Error)]
pub enum CreateReportResult {
    #[error("Success")]
    Success,
    #[error("Failed with SGX error: {}", .0.as_str())]
    Sgx(sgx_status_t),
    #[error("Failed to get Public Key")]
    FailedToGetPublicKey,
    #[error("Failed to encode Public Key")]
    FailedEncodePublicKey,
}

impl From<sgx_status_t> for CreateReportResult {
    fn from(err: sgx_status_t) -> Self {
        CreateReportResult::Sgx(err)
    }
}

pub const RSA3072_PKCS8_DER_SIZE: usize = 420;

pub const ENCLAVE_HELD_PUB_KEY_SIZE: usize = 32;

/// Size of all the enclave held data shared and validated during attestation
pub const ENCLAVE_HELD_DATA_SIZE: usize = ENCLAVE_HELD_PUB_KEY_SIZE;

pub type EnclaveHeldData = [u8; ENCLAVE_HELD_DATA_SIZE];

pub type PubkeyPkcs8 = [u8; RSA3072_PKCS8_DER_SIZE];

#[derive(Error, Copy, Clone, PartialEq, Eq, Ord, PartialOrd, Debug)]
pub enum EcallError<T: 'static + std::error::Error + Display> {
    // Error from the sgx runtime. Ecall returned with an error.
    #[error("Ecall failed in SGX runtime: {}", .0.as_str())]
    SgxRuntime(sgx_status_t),

    // Error from the RTC enclave code. This can be caused by RTC code, or a library
    // called from RTC code, including any sgx library.
    #[error("Ecall failed in the RTC enclave: {}", .0)]
    RtcEnclave(#[from] T),
}

#[cfg(test)]
mod tests {
    #[test]
    fn it_works() {
        assert_eq!(2 + 2, 4);
    }
}
