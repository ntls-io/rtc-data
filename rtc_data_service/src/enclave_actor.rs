use actix::prelude::*;
use rtc_uenclave::{AttestationError, EnclaveConfig, RtcEnclave};
use std::sync::Arc;

#[derive(Default)]
pub(crate) struct RequestAttestation;

type RequestAttestationResult = Result<String, AttestationError>;

impl Message for RequestAttestation {
    type Result = RequestAttestationResult;
}

pub(crate) struct EnclaveActor {
    enclave: Option<RtcEnclave<Arc<EnclaveConfig>>>,
    config: Arc<EnclaveConfig>,
}

impl EnclaveActor {
    pub fn new(config: Arc<EnclaveConfig>) -> Self {
        Self {
            enclave: None,
            config,
        }
    }
}

impl Drop for EnclaveActor {
    fn drop(&mut self) {
        println!("Dropping enclave actor");
    }
}

impl Actor for EnclaveActor {
    type Context = Context<EnclaveActor>;

    fn stopped(&mut self, _ctx: &mut Self::Context) {
        self.enclave.take().map(|enclave| enclave.destroy());
    }

    fn started(&mut self, _ctx: &mut Self::Context) {
        self.enclave
            .replace(RtcEnclave::init(self.config.clone()).expect("enclave to initialize"));
    }
}

impl Handler<RequestAttestation> for EnclaveActor {
    type Result = RequestAttestationResult;

    fn handle(&mut self, _msg: RequestAttestation, _ctx: &mut Self::Context) -> Self::Result {
        self.enclave
            .as_ref()
            .expect("RequestAttestation sent to uninitialized EnclaveActor")
            .dcap_attestation_azure()
    }
}

// TODO: Investigate supervisor returning `Err(Cancelled)` (see supervisor docs on Actix)
impl actix::Supervised for EnclaveActor {
    fn restarting(&mut self, _ctx: &mut Context<EnclaveActor>) {
        self.enclave
            .replace(RtcEnclave::init(self.config.clone()).expect("enclave to be initialized"))
            .map(|enc| enc.destroy());
    }
}
