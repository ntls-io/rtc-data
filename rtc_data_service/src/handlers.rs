use actix::Addr;
use actix_web::{error::ErrorInternalServerError, get, web, HttpRequest, HttpResponse};
use models::Status;

use crate::auth_enclave_actor;
use crate::auth_enclave_actor::AuthEnclaveActor;
use crate::data_enclave_actor;
use crate::data_enclave_actor::DataEnclaveActor;
use crate::merge_error::*;

pub async fn server_status(_req: HttpRequest) -> HttpResponse {
    HttpResponse::Ok().json(Status {
        status: "The server is up".to_string(),
    })
}

#[get("/auth/attest")]
pub async fn auth_enclave_attestation(
    _req: HttpRequest,
    enclave: web::Data<Addr<AuthEnclaveActor>>,
) -> actix_web::Result<String> {
    let jwt = enclave
        .send(auth_enclave_actor::RequestAttestation::default())
        .await
        .merge_err();
    dbg!(&jwt);

    match jwt {
        Ok(result) => Ok(result),
        // TODO: Look at the result here - change the error format and see if we want to sanitise the output in some way
        Err(err) => Err(ErrorInternalServerError(err)),
    }
}

#[get("/data/attest")]
pub async fn data_enclave_attestation(
    _req: HttpRequest,
    enclave: web::Data<Addr<DataEnclaveActor>>,
) -> actix_web::Result<String> {
    let jwt = enclave
        .send(data_enclave_actor::RequestAttestation::default())
        .await
        .merge_err();
    dbg!(&jwt);

    match jwt {
        Ok(result) => Ok(result),
        // TODO: Look at the result here - change the error format and see if we want to sanitise the output in some way
        Err(err) => Err(ErrorInternalServerError(err)),
    }
}

pub mod models {
    use serde::Serialize;

    #[derive(Serialize)]
    pub struct Status {
        pub status: String,
    }
}
