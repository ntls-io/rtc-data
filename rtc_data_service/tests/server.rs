use actix::Actor;
use actix_web::{
    http,
    test::{self, read_body},
    App,
};
use insta;
use rtc_data_service::enclave_actor::*;
use rtc_data_service::handlers::*;
use rtc_types::UploadMetadata;
use rtc_uenclave::EnclaveConfig;
use sgx_types::sgx_target_info_t;
use sodalite;

use std::sync::Arc;

#[actix_rt::test]
async fn data_service_attestation_ok() {
    let mut app = test::init_service(
        App::new()
            .data(
                EnclaveActor::new(Arc::new(EnclaveConfig {
                    lib_path: "/root/rtc-data/rtc_data_enclave/build/bin/enclave.signed.so"
                        .to_string(),
                    ..Default::default()
                }))
                .start(),
            )
            .service(data_enclave_attestation),
    )
    .await;

    let req = test::TestRequest::get().uri("/data/attest").to_request();
    let resp = test::call_service(&mut app, req).await;

    insta::assert_debug_snapshot!(resp);

    let body = read_body(resp).await;
    insta::assert_debug_snapshot!(body);
}

#[test]
fn test_test() {
    println!("hi");
    let enclave = rtc_uenclave::RtcEnclave::init(EnclaveConfig {
        lib_path: "/root/rtc-data/rtc_data_enclave/build/bin/enclave.signed.so".to_string(),
        ..Default::default()
    })
    .unwrap();
    let ehd = enclave
        .create_report(&sgx_target_info_t::default())
        .unwrap()
        .enclave_held_data;

    let mut pubkey = [0_u8; 32];
    let mut privkey = [0_u8; 32];

    sodalite::box_keypair_seed(&mut pubkey, &mut privkey, &[2_u8; 32]);

    let plaintext = [[0_u8; 32], [12_u8; 32]].concat();
    let mut ciphertext = vec![0_u8; plaintext.len()];
    let nonce = [8_u8; 24];

    sodalite::box_(&mut ciphertext, &plaintext, &nonce, &ehd, &privkey);

    println!("{:?}", ciphertext);

    let res = enclave.upload_data(
        &ciphertext,
        UploadMetadata {
            uploader_pub_key: pubkey,
            nonce,
        },
    );

    println!("res: {:?}", res);
}
