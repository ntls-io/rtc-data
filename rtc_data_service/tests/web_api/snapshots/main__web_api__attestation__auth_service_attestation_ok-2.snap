---
source: rtc_data_service/tests/server.rs
expression: body

---
b"Failed to get quote: SGX_QL_INTERFACE_UNAVAILABLE"
